### PR TITLE
Add stubs generator, move C++ bindings out of libsnark.ml

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* moves the ctypes bindings out of libsnark.ml
* tweaks the format of the bindings so that they interact properly with the snarky library
* creates a generator executable for the `.c` and `.ml` files

**NOTE:** Actual build is not currently using the generated stubs yet; I'm still trying to track down a double-linking bug that occurs when we try to use them:
```
...
src/camlsnark_c/libcamlsnark_c_stubs.a(common.cpp.o):src/camlsnark_c/libsnark-caml/libsnark/caml/common.cpp:90: multiple definition of `camlsnark_mnt6_final_exponent_last_chunk_abs_of_w0'
src/camlsnark_c/libcamlsnark_c_stubs.a(common.cpp.o):src/camlsnark_c/libsnark-caml/libsnark/caml/common.cpp:90: first defined here
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking
```

### CODE REVIEW

**Comment by mrmr1993** [comment]
Updated. I worked out the linking bug, and now this uses statically-linked functions for all operations on `Field.t`, etc.

**I have not tested this on a Mac.**

**Comment by mrmr1993** [comment]
I ran some profiles in coda:
* this PR
```
$ `which time` dune exec coda -- tra -witness-only -repeat 10
58.88user 1.75system 1:01.60elapsed 98%CPU (0avgtext+0avgdata 212160maxresident)k
26360inputs+8256outputs (5major+747668minor)pagefaults 0swaps
$ `which time` dune exec coda -- tra -witness-only -repeat 10
57.99user 1.51system 0:59.42elapsed 100%CPU (0avgtext+0avgdata 209668maxresident)k
0inputs+8256outputs (0major+747695minor)pagefaults 0swaps

$ (cd ..; make benchmarks); BENCHMARK_LIBRARIES=vrf_lib_tests ../scripts/benchmarks.sh
INFO Not using Docker
cd src && dune build app/benchmarks/main.exe
Running inline tests in library "vrf_lib_tests"
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌──────────────────────────────────────────────────────────────────────┬──────────┬─────────────┬────────────┬────────────┬────────────┐
│ Name                                                                 │ Time/Run │     mWd/Run │   mjWd/Run │   Prom/Run │ Percentage │
├──────────────────────────────────────────────────────────────────────┼──────────┼─────────────┼────────────┼────────────┼────────────┤
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval unchecked │   1.26ms │     72.98kw │     4.47kw │     4.17kw │      0.27% │
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval checked   │ 468.13ms │ 20_991.90kw │ 6_925.70kw │ 6_924.89kw │    100.00% │
└──────────────────────────────────────────────────────────────────────┴──────────┴─────────────┴────────────┴────────────┴────────────┘
Benchmarks that take 1ns to 100ms can be estimated precisely. For more reliable
estimates, redesign your benchmark to have a shorter execution time.
$ (cd ..; make benchmarks); BENCHMARK_LIBRARIES=vrf_lib_tests ../scripts/benchmarks.sh
INFO Not using Docker
cd src && dune build app/benchmarks/main.exe
Running inline tests in library "vrf_lib_tests"
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌──────────────────────────────────────────────────────────────────────┬──────────┬─────────────┬────────────┬────────────┬────────────┐
│ Name                                                                 │ Time/Run │     mWd/Run │   mjWd/Run │   Prom/Run │ Percentage │
├──────────────────────────────────────────────────────────────────────┼──────────┼─────────────┼────────────┼────────────┼────────────┤
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval unchecked │   1.27ms │     72.98kw │     4.47kw │     4.17kw │      0.27% │
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval checked   │ 468.68ms │ 20_991.90kw │ 6_925.70kw │ 6_924.89kw │    100.00% │
└──────────────────────────────────────────────────────────────────────┴──────────┴─────────────┴────────────┴────────────┴────────────┘
Benchmarks that take 1ns to 100ms can be estimated precisely. For more reliable
estimates, redesign your benchmark to have a shorter execution time.
```
* master
```
$ `which time` dune exec coda -- tra -witness-only -repeat 10
127.65user 1.80system 2:10.34elapsed 99%CPU (0avgtext+0avgdata 211136maxresident)k
26504inputs+8296outputs (6major+744474minor)pagefaults 0swaps
$ `which time` dune exec coda -- tra -witness-only -repeat 10
128.03user 1.64system 2:09.75elapsed 99%CPU (0avgtext+0avgdata 212136maxresident)k
0inputs+8320outputs (0major+744502minor)pagefaults 0swaps

$ (cd ..; make benchmarks); BENCHMARK_LIBRARIES=vrf_lib_tests ../scripts/benchmarks.sh
INFO Not using Docker
cd src && dune build app/benchmarks/main.exe
Running inline tests in library "vrf_lib_tests"
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌──────────────────────────────────────────────────────────────────────┬────────────┬──────────────┬────────────┬────────────┬────────────┐
│ Name                                                                 │   Time/Run │      mWd/Run │   mjWd/Run │   Prom/Run │ Percentage │
├──────────────────────────────────────────────────────────────────────┼────────────┼──────────────┼────────────┼────────────┼────────────┤
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval unchecked │     1.57ms │     181.44kw │     4.51kw │     4.21kw │      0.15% │
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval checked   │ 1_068.69ms │ 202_919.46kw │ 7_312.02kw │ 7_311.21kw │    100.00% │
└──────────────────────────────────────────────────────────────────────┴────────────┴──────────────┴────────────┴────────────┴────────────┘
Benchmarks that take 1ns to 100ms can be estimated precisely. For more reliable
estimates, redesign your benchmark to have a shorter execution time.

$ (cd ..; make benchmarks); BENCHMARK_LIBRARIES=vrf_lib_tests ../scripts/benchmarks.sh
INFO Not using Docker
cd src && dune build app/benchmarks/main.exe
Running inline tests in library "vrf_lib_tests"
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌──────────────────────────────────────────────────────────────────────┬────────────┬──────────────┬────────────┬────────────┬────────────┐
│ Name                                                                 │   Time/Run │      mWd/Run │   mjWd/Run │   Prom/Run │ Percentage │
├──────────────────────────────────────────────────────────────────────┼────────────┼──────────────┼────────────┼────────────┼────────────┤
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval unchecked │     1.59ms │     181.44kw │     4.51kw │     4.21kw │      0.15% │
│ [lib/vrf_lib/integrated_test.ml:vrf bench module] vrf eval checked   │ 1_090.99ms │ 202_919.46kw │ 7_312.02kw │ 7_311.21kw │    100.00% │
└──────────────────────────────────────────────────────────────────────┴────────────┴──────────────┴────────────┴────────────┴────────────┘
Benchmarks that take 1ns to 100ms can be estimated precisely. For more reliable
estimates, redesign your benchmark to have a shorter execution time.
```

More than 50% improvement in witness generation time. :smile:

**Comment by bkase** [comment]
Can confirm this works on macos!

**Comment by cmr** [comment]
This brings the raw `merkle_update.exe` size on linux from 81M to 72M, probably because removing `-whole-archive` lets more of libsnark get stripped away.

File: src/camlsnark_c/camlsnark_linker_flags_gen.ml [hunk 1]
```diff
@@ -17,9 +17,8 @@ let () =
         ; "-lstdc++" ]
     | "Linux" ->
         [ "-Wl,-E"
```
Comment (cmr) [inline src/camlsnark_c/camlsnark_linker_flags_gen.ml]: Can you also replace the first arg for `"Darwin"` with `-lcamlsnark_c_stubs` to get the delicious space savings?

Comment (mrmr1993) [reply src/camlsnark_c/camlsnark_linker_flags_gen.ml]: Nice, done :)

File: src/camlsnark_c/camlsnark_linker_flags_gen.ml [hunk 2]
```diff
@@ -8,7 +8,7 @@ let () =
   C.Flags.write_sexp "flags.sexp"
     ( match l with
     | "Darwin" ->
-        [ sprintf "-Wl,-force_load,%s/libcamlsnark_c_stubs.a" cwd
+        [ sprintf "-l%s/libcamlsnark_c_stubs.a" cwd
```
Comment (cmr) [inline src/camlsnark_c/camlsnark_linker_flags_gen.ml]: I think this will fail? The linker will probably barf on the full path here.

**Comment by mrmr1993** [comment]
Updated this to use the static bindings for `int`/`bool`/`long` as well. This has a much smaller impact, but we might as well do it since the infrastructure's there anyway.

```
$ `which time` dune exec coda -- tra -witness-only -repeat 10
58.28user 1.59system 0:59.81elapsed 100%CPU (0avgtext+0avgdata 214172maxresident)k
0inputs+8336outputs (0major+748056minor)pagefaults 0swaps
$ `which time` dune exec coda -- tra -witness-only -repeat 10
57.72user 1.60system 0:59.24elapsed 100%CPU (0avgtext+0avgdata 215320maxresident)k
0inputs+8336outputs (0major+747020minor)pagefaults 0swaps
```

**Pull request merged**